image/auth/env_keychain.go

```diff
@@ -1,49 +1,81 @@
 package auth
 
 import (
+	"encoding/base64"
 	"encoding/json"
+	"fmt"
 	"os"
+	"regexp"
 
 	"github.com/google/go-containerregistry/pkg/authn"
-	"github.com/google/go-containerregistry/pkg/name"
 	"github.com/pkg/errors"
 
 	"github.com/buildpack/lifecycle/cmd"
 )
 
-type EnvKeychain struct{}
-
 func DefaultEnvKeychain() authn.Keychain {
-	return authn.NewMultiKeychain(&EnvKeychain{}, authn.DefaultKeychain)
+	return authn.NewMultiKeychain(&EnvKeychain{cmd.EnvRegistryAuth}, authn.DefaultKeychain)
 }
 
-func (EnvKeychain) Resolve(registry name.Registry) (authn.Authenticator, error) {
-	env := os.Getenv(cmd.EnvRegistryAuth)
-	if env == "" {
-		return authn.Anonymous, nil
-	}
-	authMap := map[string]string{}
-	err := json.Unmarshal([]byte(env), &authMap)
+type EnvKeychain struct {
+	EnvVar string
+}
+
+func (k *EnvKeychain) Resolve(resource authn.Resource) (authn.Authenticator, error) {
+	authHeaders, err := ReadAuthEnvVar(k.EnvVar)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse %s value", cmd.EnvRegistryAuth)
+		return nil, errors.Wrap(err, "reading auth env var")
 	}
-	auth, ok := authMap[registry.Name()]
+
+	header, ok := authHeaders[resource.RegistryStr()]
 	if ok {
-		return &providedAuth{auth: auth}, nil
+		authConfig, err := authHeaderToConfig(header)
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing auth header '%s'", header)
+		}
+
+		return &providedAuth{config: authConfig}, nil
 	}
 
 	return authn.Anonymous, nil
 }
 
 type providedAuth struct {
-	auth string
+	config *authn.AuthConfig
 }
 
-func (p *providedAuth) Authorization() (string, error) {
-	return p.auth, nil
+func (p *providedAuth) Authorization() (*authn.AuthConfig, error) {
+	return p.config, nil
 }
 
-func BuildEnvVar(keychain authn.Keychain, images ...string) (string, error) {
+// ReadAuthEnvVar parses an environment variable to produce a map of 'registry url' to 'authorization header'.
+//
+// Complementary to `BuildAuthEnvVar`.
+//
+// Example Input:
+// 	{"gcr.io": "Bearer asdf=", "docker.io": "Basic qwerty="}
+//
+// Example Output:
+//  gcr.io -> Bearer asdf=
+//  docker.io -> Basic qwerty=
+func ReadAuthEnvVar(envVar string) (map[string]string, error) {
```

Comment:
Rename to `ReadEnvVar` to usage looks like `auth.ReadEnvVar` instead of `auth.ReadAuthEnvVar`